### PR TITLE
handle unversioned objects for encoding for CRs

### LINKED ISSUE #47036: CR status messages don't include kind and apiVersion
Without those fields, clients cannot deserialize the errors.

### PR DESCRIPTION
**Pull request by deads2k**
Fixes https://github.com/kubernetes/kubernetes/issues/47036

This updates the encoding for "special" types like `Status` when coming back through the CR APIs.  It also closes a bug this allowed to be exposed in validation for CRs

xref: https://github.com/kubernetes/features/issues/95

### CODE REVIEW

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-unit test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/apiserver.go [hunk 1]
```diff
@@ -56,6 +56,17 @@ var (
 	registry             = registered.NewOrDie("")
 	Scheme               = runtime.NewScheme()
 	Codecs               = serializer.NewCodecFactory(Scheme)
+
+	// if you modify this, make sure you update the crEncoder
+	unversionedVersion = schema.GroupVersion{Group: "", Version: "v1"}
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/apiserver.go]: 😫  https://github.com/kubernetes/kubernetes/pull/46112

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go [hunk 2]
```diff
@@ -353,13 +360,38 @@ func (s UnstructuredNegotiatedSerializer) SupportedMediaTypes() []runtime.Serial
...
+type crEncoder struct{}
+
+func (crEncoder) Encode(obj runtime.Object, w io.Writer) error {
+	switch obj.(type) {
+	case
+		*metav1.Status,
+		*metav1.WatchEvent:
+
+		var serializer runtime.Serializer
+		for _, info := range Codecs.SupportedMediaTypes() {
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go]: having seen this loop at a couple of places. We should have a getter for a given media type.

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go [hunk 3]
```diff
@@ -353,13 +360,38 @@ func (s UnstructuredNegotiatedSerializer) SupportedMediaTypes() []runtime.Serial
...
+			// we are always json
+			if info.MediaType == "application/json" {
+				serializer = info.Serializer
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go]: nit: `return serializer.Encode(obj, w)` here and error in the default case.

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go [hunk 4]
```diff
@@ -353,13 +360,38 @@ func (s UnstructuredNegotiatedSerializer) SupportedMediaTypes() []runtime.Serial
...
 
+var crEncoderInstance = crEncoder{}
+
+// crEncoder *usually* encodes using the unstructured.UnstructuredJSONScheme, but if the type is one of the unversioned kinds
+// it will serialize them out
+type crEncoder struct{}
+
+func (crEncoder) Encode(obj runtime.Object, w io.Writer) error {
+	switch obj.(type) {
+	case
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go]: nit: strange indention with multiple lines

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go [hunk 5]
```diff
@@ -353,13 +360,38 @@ func (s UnstructuredNegotiatedSerializer) SupportedMediaTypes() []runtime.Serial
...
+	return versioning.NewDefaultingCodecForScheme(Scheme, crEncoderInstance, nil, gv, nil)
 }
 
 func (s UnstructuredNegotiatedSerializer) DecoderToVersion(serializer runtime.Decoder, gv runtime.GroupVersioner) runtime.Decoder {
 	return unstructured.UnstructuredJSONScheme
 }
 
+var crEncoderInstance = crEncoder{}
+
+// crEncoder *usually* encodes using the unstructured.UnstructuredJSONScheme, but if the type is one of the unversioned kinds
```
Comment (sttts) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go]: actually the code is hardcoded for Status and WatchEvent, not for the unversioned types.

**Comment by sttts** [comment]
Couple of nits. Looks good in general.

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/apiserver.go (hunk 1 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/apiserver.go]: > #46112

Commented in that pull.  I'm not sure enough to merge it over clayton's objection.

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go (hunk 2 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go]: > having seen this loop at a couple of places. We should have a getter for a given media type.

I don't remember seeing it, but something in that negotiated serializer API has to change.  It's really difficult to sort out.  Feels like we're receiving information as input that we don't need.

File: staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go (hunk 3 above)
Comment (deads2k) [inline staging/src/k8s.io/kube-apiextensions-server/pkg/apiserver/customresource_handler.go]: > sttts 5 hours ago Member
> nit: return serializer.Encode(obj, w) here and error in the default case.

sure

**Comment by deads2k** [comment]
comments addressed.  Tagging.

**Comment by mikedanese** [comment]
@k8s-bot test this please

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by nikhita** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by sttts** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by deads2k** [comment]
Added issue reference, blocker for 1.7.

**Comment by deads2k** [comment]
holding for https://github.com/kubernetes/kubernetes/pull/46775

**Comment by deads2k** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Pull request merged**